service/matching/taskqueuemonitor.go

```diff
@@ -0,0 +1,120 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+)
+
+type (
+	tqmonStop    struct{}
+	tqmonGC      struct{}
+	tqmonUsurped struct{ id *taskQueueID }
+
+	// tqmonAccess is the interface by which a taskQueueMonitor observes and
+	// updates a set of taskQueueManager instances.
+	tqmonAccess interface {
+		getTaskQueues(int) []taskQueueManager
+		unloadTaskQueue(*taskQueueID)
+	}
+
+	// taskQueueMonitor performs garbage collection and unloads instances of
+	// taskQueueManager accessed via a tqmonAccess.
+	taskQueueMonitor struct {
+		sigs chan interface{}
+	}
+)
+
+// newTaskQueueMonitor constructs a new instance of taskQueueMonitor.
+func newTaskQueueMonitor() *taskQueueMonitor {
+	return &taskQueueMonitor{sigs: make(chan interface{})}
+}
+
+// Run runs the taskQueueMonitor event loop, performing GC of a set of
+// taskQueueManager instances and unloading the same when signaled or when GC
+// deems it appropriate. The implementaion of tqmonAccess MUST NOT re-enter this
+// taskQueueMonitor instance.
+func (m *taskQueueMonitor) Run(access tqmonAccess, config *Config) {
+	defer close(m.sigs)
+	t := time.NewTimer(config.IdleTaskqueueCheckInterval())
+	defer t.Stop()
+
+	for {
+		select {
+		case i := <-m.sigs:
+			switch sig := i.(type) {
+			case tqmonUsurped:
+				access.unloadTaskQueue(sig.id)
+			case tqmonGC:
+				m.runGC(access)
+				t.Reset(config.IdleTaskqueueCheckInterval())
+			case tqmonStop:
+				return
+			default:
+				panic(fmt.Sprintf(
+					"Unsupported signal type sent to taskQueueMonitor %T", sig))
+			}
+		case <-t.C:
+			m.runGC(access)
+			t.Reset(config.IdleTaskqueueCheckInterval())
+		}
+	}
+}
+
+// Stop synchronously stops a running taskQueueMonitor. Once stopped a
+// taskQueueMonitor must not be restarted. The only time this function returns
+// an error is if the provided context ends while the shutdown is in process, in
+// which case the ctx.Err() is returned and the internal state of the
+// taskQueueMonitor instance is undefined.
+func (m *taskQueueMonitor) Stop(ctx context.Context) error {
+	select {
+	case m.sigs <- tqmonStop{}:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	select {
+	case <-m.sigs:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return nil
+}
+
+// Signal sends the supplied message to the taskQueueMonitor. Supported message
+// types are tqmonStop, tqmonUsurped, and tqmonGC
+func (m *taskQueueMonitor) Signal(i interface{}) {
```

Comment:
I feel like all these signal types should be only internal to taskQueueMonitor and user of tqMonitor does not need to know them. taskQueueMonitor could define functions that wrap those signals. Also, is tqmonGC only needed for tests? I don't see where it is used in non-test code.

Reply:
Yeah, there are tests that depend on changing the GC interval on the fly.

Reply:
Can you achieve the same with a mock clock and forward the time when needed?

Reply:
Not with our in-house timesource mock. The only lib I'm aware of that could do what you describe is https://github.com/benbjohnson/clock. Last time I looked at it (years ago) there were problems but it may be better now.

Reply:
Pretty sure this was the problem back then so there's a good chance it's fixed now. https://github.com/benbjohnson/clock/commit/6b1ff0d3224577b9e32ee68cce3813c42af92f32